### PR TITLE
Install configuration upgrade script and improve testing

### PR DESCRIPTION
**Pull request by ansasaki**
The main changes included in this patch set are the installation of the `convert_config.py` script as `keylime_upgrade_config` alongside other binaries and changes to the tests to make them to not install any binaries nor configuration files during execution.

The `keylime_upgrade_config` will use the `/etc/keylime` as the default output location when executed as root, otherwise it will use the current working directory as the output directory. The `keylime_upgrade_config` only performs upgrades when there is at least one configuration file that is not up-to-date. It compares each configuration version with the versions available in the templates directory (by default `/usr/share/keylime/templates`).

During startup of agent, verifier, registrar, or tenant, the respective configuration file version is compared with the latest upgrade template version available and will log a message in case an upgrade is available (it will not perform the upgrade automatically). When a minor upgrade is available, a message with INFO level is logged informing the available upgrade. If a major upgrade is available, a message with WARNING level is logged.

The configuration processing was modified to remove the quotes from strings obtained from the configuration files. This improves the compatibility with the TOML format used for the rust agent.

The tests were modified to not install keylime binaries nor configuration files during execution (the dependencies are still installed). A virtualenv and temporary files are used instead.

Superseds #1132

### CODE REVIEW

File: keylime/cmd/convert_config.py
Comment (aplanas) [inline keylime/cmd/convert_config.py:372]: why is this needed? is not the configparser behaving like a dict of dicts?

**Review by aplanas — COMMENTED** [review]
[Continue later]

File: .gitignore
Comment (ansasaki) [reply .gitignore:40]: Without this gitignore considers `config/` as a matching pattern to exclude, which means it will exclude any directory named `config` and not only the one in the git root.

By adding `/` before config, it will only exclude the `config` directory at the root and not `test/data/config` for example.

I mentioned this in the commit message. Please refer to the documentation for further information:
https://git-scm.com/docs/gitignore

Comment (aplanas) [reply .gitignore:40]: Yes, thanks!

File: MANIFEST.in
Comment (ansasaki) [reply MANIFEST.in:3]: To recursively  include the `templates` directory during installation. I'm not a python expert, this was the easiest way I found.

File: keylime/cmd/convert_config.py
Comment (ansasaki) [reply keylime/cmd/convert_config.py:372]: `TypeError: Object of type RawConfigParser is not JSON serializable`

File: keylime/cmd/convert_config.py
Comment (ansasaki) [reply keylime/cmd/convert_config.py:283]: TBH I don't know what are the pros and cons here, I just used what we were already using.

File: keylime/cmd/convert_config.py
Comment (ansasaki) [reply keylime/cmd/convert_config.py:285]: Thanks, I will drop it

File: keylime/cmd/convert_config.py
Comment (ansasaki) [reply keylime/cmd/convert_config.py:288]: Thanks!

File: keylime/cmd/convert_config.py
Comment (ansasaki) [reply keylime/cmd/convert_config.py:114]: Interesting, I think it makes easier to read. I'll replace it

File: templates/2.0/adjust.py
Comment (aplanas) [inline templates/2.0/adjust.py:141]: `config[component][option] = str(list(v))`

File: templates/2.0/adjust.py
Comment (aplanas) [inline templates/2.0/adjust.py:158]: `str(v)`

File: templates/2.0/logging.j2
Comment (aplanas) [inline templates/2.0/logging.j2:5]: Do we expect quotes here too and in the rest?

File: templates/2.0/mapping.json
Comment (aplanas) [inline templates/2.0/mapping.json:3]: too late to drop this level too?

File: templates/2.0/mapping.json
Comment (aplanas) [inline templates/2.0/mapping.json:793]: Ah. Forget all what I said about the first level. Sorry for the noise.

**Review by aplanas — COMMENTED** [review]
Thanks for this, that was a lot of work!

File: pyproject.toml
Comment (ansasaki) [reply pyproject.toml:13]: I could not find any way to do this via code annotations. The problem here is that the file is intentionally not parseable, but the annotations are evaluated after code parsing, which leads to errors.

The only way I found to please the linters was to skip the file completely. For black the options were to change the configuration file or to explicitly exclude on command line with `--exclude`.

File: keylime/config.py
Comment (ansasaki) [reply keylime/config.py:371]: Right, we can avoid iterating here, thanks!

File: keylime/config.py
Comment (ansasaki) [reply keylime/config.py:364]: Hum, I think now it can. Previously, the code was mixing `dict` and `RawConfigParser` which made `get()` to behave differently.  Now, the code uses only `RawConfigParser` which makes `get()` to be consistent.

File: templates/2.0/adjust.py
Comment (ansasaki) [reply templates/2.0/adjust.py:8]: I agree it makes sense to move this to a module to avoid duplication. Will add a commit for that.

File: templates/2.0/adjust.py
Comment (ansasaki) [reply templates/2.0/adjust.py:94]: Yes, I agree. I'll rename it to `section`

File: templates/2.0/adjust.py
Comment (ansasaki) [reply templates/2.0/adjust.py:57]: Since this adjustment only applies to the `agent`, I think it is better to drop the dict and use a list.

File: templates/2.0/adjust.py
Comment (ansasaki) [reply templates/2.0/adjust.py]: Done

File: templates/2.0/adjust.py
Comment (ansasaki) [reply templates/2.0/adjust.py]: Done

File: templates/2.0/adjust.py
Comment (ansasaki) [reply templates/2.0/adjust.py]: Done

File: templates/2.0/adjust.py
Comment (ansasaki) [reply templates/2.0/adjust.py]: Done

File: templates/2.0/adjust.py
Comment (ansasaki) [reply templates/2.0/adjust.py]: Done

File: keylime/config.py
Comment (ansasaki) [reply keylime/config.py]: Done

File: pyproject.toml
Comment (ansasaki) [reply pyproject.toml:13]: It seems that the command in CI is ignoring the configuration file (`pyproject.toml`) which makes the file `test/data/template-invalid-adjust/2.0/adjust.py` to not be skipped, causing error. (sigh)

I'll try to fix this by modifying the command line. If I don't succeed, I'll simply drop the test case I wrote.

Comment (ansasaki) [reply pyproject.toml:13]: Ok, it seems I managed to make `pre-commit` to ignore the intentionally invalid file!

**Comment by ansasaki** [comment]
I could not re-run the PR code coverage measurement check. After the test changes (https://github.com/RedHat-SP-Security/keylime-tests/pull/284) are merged, and I rebase to drop the temporary commit that changes the tests reference, I believe it will re-run without issues.

File: installer.sh
Comment (THS-on) [inline installer.sh:440]: We can also store them inside the package: https://setuptools.pypa.io/en/stable/userguide/datafiles.html

The advantage of that approach is that it the Python way, the disadvantage is that distributions normally would expect those files in `/usr/share`

**Review by THS-on — COMMENTED** [review]
In general LGTM. I have only a question regarding where to store the templates.

File: installer.sh
Comment (ansasaki) [reply installer.sh:440]: I'm not a python expert, but I thought the changes I made already included the templates as part of the python distribution package. Please check, and let me know what I need to change to make it to be included.

File: keylime/cmd/convert_config.py
Comment (THS-on) [reply keylime/cmd/convert_config.py:274]: I think I switched to RawConfigParser to be able to pass the config to the logging module. I don't know if this is relevant in this case.

File: installer.sh
Comment (THS-on) [reply installer.sh:440]: You are right that the files are included. I missed that, sorry. 

Currently we default to use the files `/usr/share/keylime/`, I would instead load them from the package by default (https://docs.python.org/3/library/importlib.resources.html).

**Review by aplanas — APPROVED** [review]
I guess that the last commit will be dropped, but LGTM!

File: installer.sh
Comment (ansasaki) [reply installer.sh:440]: I will try to use the templates from the package  (when available) as an alternative location when `/usr/share/keylime/templates` is not available. Would this be OK for you?

**Comment by ansasaki** [comment]
> I guess that the last commit will be dropped, but LGTM!

Yes, the last commit will be dropped before merge. But it is blocked by  https://github.com/RedHat-SP-Security/keylime-tests/pull/284

File: installer.sh
Comment (THS-on) [reply installer.sh:440]: Yes, that sounds like a good idea!

Comment (ansasaki) [reply installer.sh:440]: So, after reading some documentation, I learned that to use the templates from the package as you suggested
would require changes in many places including changing the whole upgrade process.

The resources are not supposed to be treated as files in the file system, and I use subdirectories to identify the available versions. This means that the logic how the available versions are handled would be different. That would require changes to tests as well.

We would also need to move the `templates` directory to be inside the keylime package (i.e.`keylime/templates`), as well as installing them as part of the binary distribution. The templates would have to become a subpackage (i.e. `import keylime.templates` works).

For me, this looks like a lot of work that would complicate the code and maintainability (as we would have two separate ways of handling the files and detecting the available versions to be tested and maintained), with little gain.

If this is really required, I would work on it on a separate PR.

Comment (THS-on) [reply installer.sh:440]: Yeah, that sounds more complicated than I thought. Putting those files into `/usr/share` is standard for Linux packaging, so I'm fine with using that!

**Comment by THS-on** [comment]
I just noticed that if you run it on an older installation of Keylime the upgrade script runs again because of the switch to quoting strings. E.g.:
```diff
 # The configuration file version number
-version = 2.0
+version = "2.0"
```
Is this intentional? In this case running the migration is fine, but it might be a little bit  confusing.


Otherwise this change LGTM.

**Comment by ansasaki** [comment]
> I just noticed that if you run it on an older installation of Keylime the upgrade script runs again because of the switch to quoting strings. E.g.:
> 
> ```diff
>  # The configuration file version number
> -version = 2.0
> +version = "2.0"
> ```
> 
> Is this intentional? In this case running the migration is fine, but it might be a little bit confusing.

It is not :). Thanks for reporting this, I'll fix the bug

**Comment by ansasaki** [comment]
The linter issue seems to be addressed by #1267

**Comment by ansasaki** [comment]
The failure in Fedora 36 is related with a regression introduced by an update from tpm2-tss from version 3.2.0 to 3.2.1, which was reported upstream. The `tpm2_getekcertificate` command cannot retrieve the EK certificate on the affected version.

The failure is unrelated with the changes introduced by this PR. After dropping the commit that changed the used tests branch this is ready to be merged.

See:
https://github.com/tpm2-software/tpm2-tss/issues/2537
https://bugzilla.redhat.com/show_bug.cgi?id=2158598

**Review by THS-on — APPROVED** [review]
LGTM

**Pull request merged**
